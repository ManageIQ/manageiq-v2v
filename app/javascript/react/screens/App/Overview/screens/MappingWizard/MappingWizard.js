import React from 'react';
import PropTypes from 'prop-types';
import {
  bindMethods,
  noop,
  Button,
  Icon,
  Modal,
  Wizard
} from 'patternfly-react';
import {
  createTransformationMappings,
  getMappedSourceClusters,
  getSourceClustersWithMappings
} from './helpers';
import MappingWizardBody from './MappingWizardBody';
import WarningModal from './components/WarningModal/WarningModal';

const mappingWizardSteps = [
  'mappingWizardGeneralStep',
  'mappingWizardClustersStep',
  'mappingWizardDatastoresStep',
  'mappingWizardNetworksStep',
  'mappingWizardResultsStep'
];

class MappingWizard extends React.Component {
  constructor() {
    super();
    this.state = { activeStepIndex: 0 };
    bindMethods(this, ['prevStep', 'nextStep', 'goToStep']);
  }

  prevStep() {
    const { activeStepIndex } = this.state;
    this.setState({ activeStepIndex: Math.max(activeStepIndex - 1, 0) });
  }

  nextStep() {
    const { activeStepIndex } = this.state;
    const {
      mappingWizardGeneralStep,
      mappingWizardClustersStep,
      mappingWizardDatastoresStep,
      mappingWizardNetworksStep,
      setTransformationsBodyAction,
      showWarningModalAction,
      hideWarningModalAction,
      warningModalVisible
    } = this.props;

    if (activeStepIndex === 2 && !warningModalVisible) {
      const { clusterMappings } = mappingWizardClustersStep.values;
      const { datastoresMappings } = mappingWizardDatastoresStep.values;

      const sourceClustersWithoutDatastoresMappings = getMappedSourceClusters(
        clusterMappings
      ).filter(
        sourceCluster =>
          !getSourceClustersWithMappings(datastoresMappings).includes(
            sourceCluster.id
          )
      );

      if (sourceClustersWithoutDatastoresMappings.length > 0) {
        showWarningModalAction(sourceClustersWithoutDatastoresMappings);
      } else {
        this.setState({
          activeStepIndex: Math.min(
            activeStepIndex + 1,
            mappingWizardSteps.length - 1
          )
        });
      }
    } else if (activeStepIndex === 3 && !warningModalVisible) {
      const { clusterMappings } = mappingWizardClustersStep.values;
      const { networksMappings } = mappingWizardNetworksStep.values;

      const sourceClustersWithoutNetworksMappings = getMappedSourceClusters(
        clusterMappings
      ).filter(
        sourceCluster =>
          !getSourceClustersWithMappings(networksMappings).includes(
            sourceCluster.id
          )
      );

      if (sourceClustersWithoutNetworksMappings.length > 0) {
        showWarningModalAction(sourceClustersWithoutNetworksMappings);
      } else {
        const transformationsBody = createTransformationMappings(
          mappingWizardGeneralStep,
          mappingWizardClustersStep,
          mappingWizardDatastoresStep,
          mappingWizardNetworksStep
        );
        setTransformationsBodyAction(transformationsBody);
        this.setState({
          activeStepIndex: Math.min(
            activeStepIndex + 1,
            mappingWizardSteps.length - 1
          )
        });
      }
    } else {
      // Either we are not on the datastores or networks step, or we are clicking
      // the continue button in the warning modal
      hideWarningModalAction();
      if (activeStepIndex === 3) {
        const transformationsBody = createTransformationMappings(
          mappingWizardGeneralStep,
          mappingWizardClustersStep,
          mappingWizardDatastoresStep,
          mappingWizardNetworksStep
        );
        setTransformationsBodyAction(transformationsBody);
      }
      this.setState({
        activeStepIndex: Math.min(
          activeStepIndex + 1,
          mappingWizardSteps.length - 1
        )
      });
    }
  }

  goToStep(activeStepIndex) {
    const { activeStepIndex: currentStep } = this.state;

    if (currentStep === 2 && activeStepIndex === 3) {
      this.nextStep();
    } else {
      this.setState(() => ({ activeStepIndex }));
    }
  }

  render() {
    const {
      hideMappingWizard,
      hideMappingWizardAction,
      mappingWizardExitedAction,
      hideWarningModalAction,
      warningModalVisible,
      sourceClustersWithoutMappings,
      mappingWizardGeneralStep,
      mappingWizardClustersStep,
      mappingWizardDatastoresStep,
      mappingWizardNetworksStep,
      alertText,
      alertType,
      hideAlertAction
    } = this.props;

    const { activeStepIndex, transformationsBody } = this.state;
    const activeStep = (activeStepIndex + 1).toString();
    const onFirstStep = activeStepIndex === 0;
    const onFinalStep = activeStepIndex === mappingWizardSteps.length - 1;

    const currentStepProp = !onFinalStep && mappingWizardSteps[activeStepIndex];
    const currentStepForm = !onFinalStep && this.props[currentStepProp];
    const disableNextStep = !onFinalStep && !!currentStepForm.syncErrors;

    return (
      <Modal
        show={!hideMappingWizard}
        onHide={() => {
          hideMappingWizardAction(onFinalStep);
        }}
        onExited={mappingWizardExitedAction}
        dialogClassName="modal-lg wizard-pf"
      >
        <Wizard>
          <Modal.Header>
            <button
              className="close"
              onClick={() => {
                hideMappingWizardAction(onFinalStep);
              }}
              aria-hidden="true"
              aria-label="Close"
            >
              <Icon type="pf" name="close" />
            </button>
            <Modal.Title>{__('Infrastructure Mapping Wizard')}</Modal.Title>
          </Modal.Header>
          <Modal.Body className="wizard-pf-body clearfix">
            <MappingWizardBody
              loaded
              activeStepIndex={activeStepIndex}
              activeStep={activeStep}
              goToStep={this.goToStep}
              disableNextStep={disableNextStep}
              transformationsBody={transformationsBody}
              mappingWizardGeneralStep={mappingWizardGeneralStep}
              mappingWizardClustersStep={mappingWizardClustersStep}
              mappingWizardDatastoresStep={mappingWizardDatastoresStep}
              mappingWizardNetworksStep={mappingWizardNetworksStep}
              alertText={alertText}
              alertType={alertType}
              hideAlertAction={hideAlertAction}
            />
          </Modal.Body>
          <Modal.Footer className="wizard-pf-footer">
            <Button
              bsStyle="default"
              className="btn-cancel"
              onClick={() => {
                hideMappingWizardAction(onFinalStep);
              }}
              disabled={onFinalStep}
            >
              {__('Cancel')}
            </Button>

            <Button
              bsStyle="default"
              onClick={this.prevStep}
              disabled={onFirstStep || onFinalStep}
            >
              <Icon type="fa" name="angle-left" />
              {__('Back')}
            </Button>
            <Button
              bsStyle="primary"
              onClick={
                onFinalStep
                  ? () => hideMappingWizardAction(onFinalStep)
                  : this.nextStep
              }
              disabled={disableNextStep}
            >
              {onFinalStep
                ? __('Close')
                : activeStepIndex === 3 ? __('Create') : __('Next')}
              <Icon type="fa" name="angle-right" />
            </Button>
          </Modal.Footer>
        </Wizard>
        <WarningModal
          warningModalVisible={warningModalVisible}
          hideWarningModalAction={hideWarningModalAction}
          onFinalStep={onFinalStep}
          activeStepIndex={activeStepIndex}
          nextStep={this.nextStep}
          sourceClustersWithoutMappings={sourceClustersWithoutMappings}
        />
      </Modal>
    );
  }
}
MappingWizard.propTypes = {
  hideMappingWizard: PropTypes.bool,
  hideMappingWizardAction: PropTypes.func,
  mappingWizardExitedAction: PropTypes.func,
  mappingWizardGeneralStep: PropTypes.object,
  mappingWizardClustersStep: PropTypes.object,
  mappingWizardDatastoresStep: PropTypes.object,
  mappingWizardNetworksStep: PropTypes.object,
  setTransformationsBodyAction: PropTypes.func,
  showWarningModalAction: PropTypes.func,
  hideWarningModalAction: PropTypes.func,
  warningModalVisible: PropTypes.bool,
  sourceClustersWithoutMappings: PropTypes.array,
  alertText: PropTypes.string,
  alertType: PropTypes.string,
  hideAlertAction: PropTypes.func
};
MappingWizard.defaultProps = {
  hideMappingWizard: true,
  hideMappingWizardAction: noop,
  mappingWizardExitedAction: noop,
  mappingWizardGeneralStep: {},
  mappingWizardClustersStep: {},
  mappingWizardDatastoresStep: {},
  mappingWizardNetworksStep: {},
  setTransformationsBodyAction: noop,
  hideAlertAction: noop
};
export default MappingWizard;
