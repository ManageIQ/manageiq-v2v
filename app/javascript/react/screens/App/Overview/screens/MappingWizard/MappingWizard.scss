@import './components/DualPaneMapper/DualPaneMapper.scss';
@import './components/SourceClusterSelect/SourceClusterSelect.scss';
@import './components/WarningModal/WarningModal.scss';

.dual-pane-mapper-form {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.dual-pane-mapper-form.is-hidden {
  opacity: 0;
}

// TreeView container styles
.treeview-container {
  margin-top: 20px;
  width: 450px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.treeview-container--body {
  border: 1px solid rgba(0, 0, 0, 0.3);
  margin-bottom: 10px;
  padding: 10px 5px;
  height: 200px;
  width: 100%;
  overflow-y: auto;
}

.treeview__is-empty {
  text-align: center;
}

// Override bootstrap-treeview styles
.treeview span.indent {
  margin-left: 0;
}

.wizard-pf-main .blank-slate-pf {
  padding: 30px;
}

.wizard-pf-steps .wizard-pf-steps-indicator li.is-disabled a {
  cursor: not-allowed;
}

.wizard-pf-steps.step-buttons-disabled .wizard-pf-steps-indicator li.wizard-pf-step a {
  cursor: default;
}

.wizard-pf-steps.step-buttons-disabled .wizard-pf-steps-indicator li a:hover .wizard-pf-step-number {
  background-color: #fff;
  color: #bbb;
  border-color: #bbb;
}

.wizard-pf-steps.step-buttons-disabled .wizard-pf-steps-indicator li.active a:hover .wizard-pf-step-number {
  background-color: #39a5dc;
  border-color: #39a5dc;
  color: #fff;
}