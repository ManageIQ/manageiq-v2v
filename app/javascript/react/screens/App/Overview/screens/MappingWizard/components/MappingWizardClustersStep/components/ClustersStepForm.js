import React from 'react';
import PropTypes from 'prop-types';
import { bindMethods } from 'patternfly-react';

import DualPaneMapper from '../../DualPaneMapper/DualPaneMapper';
import DualPaneMapperList from '../../DualPaneMapper/DualPaneMapperList';
import DualPaneMapperCount from '../../DualPaneMapper/DualPaneMapperCount';
import DualPaneMapperListItem from '../../DualPaneMapper/DualPaneMapperListItem';
import ClustersStepTreeView from './ClustersStepTreeView';

import { sourceClustersFilter } from '../MappingWizardClustersStepSelectors';

class ClustersStepForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedTargetCluster: null,
      selectedSourceClusters: [],
      selectedMapping: null
    };

    bindMethods(this, [
      'selectSourceCluster',
      'selectTargetCluster',
      'addMapping',
      'selectMapping',
      'removeMapping',
      'removeAll'
    ]);
  }

  selectSourceCluster(sourceCluster) {
    this.setState(prevState => {
      const isAlreadySelected = prevState.selectedSourceClusters.some(
        cluster => cluster.id === sourceCluster.id
      );
      if (isAlreadySelected) {
        return {
          selectedSourceClusters: prevState.selectedSourceClusters.filter(
            cluster => cluster.id !== sourceCluster.id
          )
        };
      }
      return {
        selectedSourceClusters: [
          ...prevState.selectedSourceClusters,
          sourceCluster
        ]
      };
    });
  }

  selectTargetCluster(targetCluster) {
    this.setState(() => ({ selectedTargetCluster: targetCluster }));
  }

  addMapping() {
    const { input } = this.props;
    this.setState(prevState => {
      const clustersStepMappings = input.value;

      const mappingExistsForSelectedTargetCluster = clustersStepMappings.some(
        targetClusterWithSourceClusters =>
          targetClusterWithSourceClusters.id ===
          prevState.selectedTargetCluster.id
      );

      if (mappingExistsForSelectedTargetCluster) {
        const updatedMappings = clustersStepMappings.map(
          targetClusterWithSourceClusters => {
            if (
              targetClusterWithSourceClusters.id ===
              prevState.selectedTargetCluster.id
            ) {
              return {
                ...targetClusterWithSourceClusters,
                nodes: targetClusterWithSourceClusters.nodes.concat(
                  prevState.selectedSourceClusters.map(sourceCluster => ({
                    ...sourceCluster,
                    text: sourceCluster.name,
                    icon: 'fa fa-file-o'
                  }))
                )
              };
            }
            return targetClusterWithSourceClusters;
          }
        );
        input.onChange(updatedMappings);
      } else {
        const targetClusterWithSourceClusters = {
          ...prevState.selectedTargetCluster,
          text: prevState.selectedTargetCluster.name,
          state: {
            expanded: true
          },
          selectable: true,
          selected: false,
          nodes: prevState.selectedSourceClusters.map(sourceCluster => ({
            ...sourceCluster,
            text: sourceCluster.name,
            icon: 'fa fa-file-o'
          }))
        };
        input.onChange([
          ...clustersStepMappings,
          targetClusterWithSourceClusters
        ]);
      }
      return {
        selectedTargetCluster: null,
        selectedSourceClusters: []
      };
    });
  }

  selectMapping(selectedMapping) {
    const { input } = this.props;
    input.onChange(
      input.value.map(mapping => {
        if (mapping.id === selectedMapping.id) {
          return { ...mapping, selected: !mapping.selected };
        } else if (mapping.id !== selectedMapping.id && mapping.selected) {
          return { ...mapping, selected: false };
        }
        return mapping;
      })
    );
    this.setState(() => ({ selectedMapping }));
  }

  removeMapping() {
    const { input } = this.props;
    this.setState(prevState => {
      const clustersStepMappings = input.value;
      input.onChange(
        clustersStepMappings.filter(
          targetClusterWithSourceClusters =>
            !(
              targetClusterWithSourceClusters.id ===
              prevState.selectedMapping.id
            )
        )
      );
      return {
        selectedMapping: null
      };
    });
  }

  removeAll() {
    const { input } = this.props;
    input.onChange([]);
  }

  render() {
    const {
      sourceClusters,
      targetClusters,
      isFetchingSourceClusters,
      isFetchingTargetClusters,
      input
    } = this.props;

    const {
      selectedTargetCluster,
      selectedSourceClusters,
      selectedMapping
    } = this.state;

    return (
      <div className="dual-pane-mapper-form">
        <DualPaneMapper
          handleButtonClick={this.addMapping}
          validMapping={
            !(
              selectedTargetCluster &&
              (selectedSourceClusters && selectedSourceClusters.length > 0)
            )
          }
        >
          {sourceClusters && (
            <DualPaneMapperList
              listTitle="Source Clusters"
              loading={isFetchingSourceClusters}
            >
              {sourceClustersFilter(sourceClusters, input.value).map(item => (
                <DualPaneMapperListItem
                  item={item}
                  key={item.id}
                  selected={
                    selectedSourceClusters &&
                    selectedSourceClusters.some(
                      sourceCluster => sourceCluster.id === item.id
                    )
                  }
                  handleClick={this.selectSourceCluster}
                  handleKeyPress={this.selectSourceCluster}
                />
              ))}
              <DualPaneMapperCount
                selectedItems={selectedSourceClusters.length}
                totalItems={
                  sourceClustersFilter(sourceClusters, input.value).length
                }
              />
            </DualPaneMapperList>
          )}
          {targetClusters && (
            <DualPaneMapperList
              listTitle="Target Clusters"
              loading={isFetchingTargetClusters}
            >
              {targetClusters.map(item => (
                <DualPaneMapperListItem
                  item={item}
                  key={item.id}
                  selected={
                    selectedTargetCluster &&
                    selectedTargetCluster.id === item.id
                  }
                  handleClick={this.selectTargetCluster}
                  handleKeyPress={this.selectTargetCluster}
                />
              ))}
            </DualPaneMapperList>
          )}
        </DualPaneMapper>
        {input.value && (
          <ClustersStepTreeView
            mappings={input.value}
            selectMapping={this.selectMapping}
            removeMapping={this.removeMapping}
            removeAll={this.removeAll}
            selectedMapping={selectedMapping}
          />
        )}
      </div>
    );
  }
}

ClustersStepForm.propTypes = {
  input: PropTypes.object,
  sourceClusters: PropTypes.arrayOf(PropTypes.object),
  targetClusters: PropTypes.arrayOf(PropTypes.object),
  isFetchingSourceClusters: PropTypes.bool,
  isFetchingTargetClusters: PropTypes.bool
};

export default ClustersStepForm;
