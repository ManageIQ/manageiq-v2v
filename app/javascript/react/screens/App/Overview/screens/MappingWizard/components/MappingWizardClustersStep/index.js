import { connect } from 'react-redux';
import MappingWizardClustersStep from './MappingWizardClustersStep';
import * as MappingWizardClustersStepActions from './MappingWizardClustersStepActions';

import reducer from './MappingWizardClustersStepReducer';

export const reducers = { mappingWizardClustersStep: reducer };

const mapStateToProps = ({ mappingWizardClustersStep }, ownProps) => ({
  ...mappingWizardClustersStep,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(
  mapStateToProps,
  MappingWizardClustersStepActions,
  mergeProps
)(MappingWizardClustersStep);
