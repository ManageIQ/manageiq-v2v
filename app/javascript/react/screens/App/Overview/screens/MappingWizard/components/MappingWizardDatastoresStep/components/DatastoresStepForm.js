import React from 'react';
import PropTypes from 'prop-types';
import { bindMethods } from 'patternfly-react';
import cx from 'classnames';

import DualPaneMapper from '../../DualPaneMapper/DualPaneMapper';
import DualPaneMapperList from '../../DualPaneMapper/DualPaneMapperList';
import DualPaneMapperCount from '../../DualPaneMapper/DualPaneMapperCount';
import DualPaneMapperListItem from '../../DualPaneMapper/DualPaneMapperListItem';
import MappingWizardTreeView from '../../MappingWizardTreeView/MappingWizardTreeView';

import { sourceDatastoreFilter } from '../MappingWizardDatastoresStepSelectors';
import {
  diskSpaceInfo,
  sourceDatastoreInfo,
  targetDatastoreInfo
} from './helpers';

class DatastoresStepForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedSourceDatastores: [],
      selectedTargetDatastore: null,
      selectedMapping: null
    };

    bindMethods(this, [
      'selectSourceDatastore',
      'selectTargetDatastore',
      'addDatastoreMapping',
      'selectMapping',
      'removeMapping',
      'removeAll'
    ]);
  }

  componentWillReceiveProps(nextProps) {
    if (
      this.props.selectedCluster &&
      nextProps.selectedCluster.id !== this.props.selectedCluster.id
    ) {
      this.setState(() => ({
        selectedSourceDatastores: [],
        selectedTargetDatastore: null
      }));
    }
  }

  selectSourceDatastore(sourceDatastore) {
    this.setState(prevState => {
      const isAlreadySelected = prevState.selectedSourceDatastores.some(
        datastore => datastore.id === sourceDatastore.id
      );
      if (isAlreadySelected) {
        return {
          selectedSourceDatastores: prevState.selectedSourceDatastores.filter(
            datastore => datastore.id !== sourceDatastore.id
          )
        };
      }
      return {
        selectedSourceDatastores: [
          ...prevState.selectedSourceDatastores,
          sourceDatastore
        ]
      };
    });
  }

  selectTargetDatastore(targetDatastore) {
    this.setState(() => ({ selectedTargetDatastore: targetDatastore }));
  }

  addDatastoreMapping() {
    const { input, selectedCluster, selectedClusterMapping } = this.props;

    const { selectedTargetDatastore } = this.state;

    const noMappingForTargetCluster = !input.value.some(
      datastoreMapping => datastoreMapping.id === selectedClusterMapping.id
    );

    const addingToExistingMapping = input.value.some(
      targetClusterDatastoreMappings =>
        targetClusterDatastoreMappings.nodes.some(
          datastoreMapping => datastoreMapping.id === selectedTargetDatastore.id
        )
    );

    this.setState(prevState => {
      if (input.value.length === 0 || noMappingForTargetCluster) {
        input.onChange([
          ...input.value,
          {
            ...selectedClusterMapping,
            text: selectedClusterMapping.name,
            selectable: false,
            nodes: [
              {
                ...prevState.selectedTargetDatastore,
                text: diskSpaceInfo(
                  prevState.selectedTargetDatastore,
                  prevState.selectedSourceDatastores
                ),
                selectable: true,
                selected: false,
                state: {
                  expanded: true
                },
                nodes: prevState.selectedSourceDatastores.map(datastore => ({
                  ...datastore,
                  text: datastore.name,
                  icon: 'fa fa-file-o',
                  sourceClusterId: selectedCluster.id
                }))
              }
            ]
          }
        ]);
      } else {
        input.onChange(
          input.value.map(datastoreMapping => {
            if (datastoreMapping.id !== selectedClusterMapping.id) {
              return datastoreMapping;
            } else if (addingToExistingMapping) {
              return {
                ...datastoreMapping,
                nodes: datastoreMapping.nodes.map(mapping => {
                  if (mapping.id === prevState.selectedTargetDatastore.id) {
                    return {
                      ...mapping,
                      text: diskSpaceInfo(
                        mapping,
                        mapping.nodes.concat(prevState.selectedSourceDatastores)
                      ),
                      nodes: mapping.nodes.concat(
                        prevState.selectedSourceDatastores.map(datastore => ({
                          ...datastore,
                          text: datastore.name,
                          icon: 'fa fa-file-o',
                          sourceClusterId: selectedCluster.id
                        }))
                      )
                    };
                  }
                  return mapping;
                })
              };
            }
            return {
              ...datastoreMapping,
              nodes: datastoreMapping.nodes.concat({
                ...prevState.selectedTargetDatastore,
                text: diskSpaceInfo(
                  prevState.selectedTargetDatastore,
                  prevState.selectedSourceDatastores
                ),
                selectable: true,
                selected: false,
                state: {
                  expanded: true
                },
                nodes: prevState.selectedSourceDatastores.map(datastore => ({
                  ...datastore,
                  text: datastore.name,
                  icon: 'fa fa-file-o',
                  sourceClusterId: selectedCluster.id
                }))
              })
            };
          })
        );
      }
      return {
        selectedTargetDatastore: null,
        selectedSourceDatastores: []
      };
    });
  }

  selectMapping(selectedMapping) {
    const { input } = this.props;

    input.onChange(
      input.value.map(targetClusterDatastoreMappings => {
        const updatedMappings = targetClusterDatastoreMappings.nodes.map(
          datastoreMapping => {
            if (datastoreMapping.id === selectedMapping.id) {
              return {
                ...datastoreMapping,
                selected: !datastoreMapping.selected
              };
            } else if (
              datastoreMapping.id !== selectedMapping.id &&
              datastoreMapping.selected
            ) {
              return { ...datastoreMapping, selected: false };
            }
            return datastoreMapping;
          }
        );
        return { ...targetClusterDatastoreMappings, nodes: updatedMappings };
      })
    );
    this.setState(() => ({ selectedMapping }));
  }

  removeMapping() {
    const { input } = this.props;
    const { selectedMapping } = this.state;
    const datastoresStepMappings = input.value;

    const updatedDatastoresStepMappings = datastoresStepMappings
      .map(targetClusterWithDatastoreMappings => {
        const containsMappingToRemove = targetClusterWithDatastoreMappings.nodes.some(
          mapping => mapping.id === selectedMapping.id
        );
        if (!containsMappingToRemove) {
          return targetClusterWithDatastoreMappings;
        }
        const updatedDatastoreMappings = targetClusterWithDatastoreMappings.nodes.filter(
          datastoreMapping => !(datastoreMapping.id === selectedMapping.id)
        );
        if (updatedDatastoreMappings.length === 0) {
          return undefined;
        }
        return {
          ...targetClusterWithDatastoreMappings,
          nodes: updatedDatastoreMappings
        };
      })
      .filter(item => item !== undefined);

    input.onChange(updatedDatastoresStepMappings);
    this.setState(() => ({ selectedMapping: null }));
  }

  removeAll() {
    const { input } = this.props;
    input.onChange([]);
  }

  render() {
    const {
      sourceDatastores,
      targetDatastores,
      isFetchingSourceDatastores,
      isFetchingTargetDatastores,
      input,
      selectedCluster
    } = this.props;
    const {
      selectedSourceDatastores,
      selectedTargetDatastore,
      selectedMapping
    } = this.state;

    const classes = cx('dual-pane-mapper-form', {
      'is-hidden': !selectedCluster
    });

    return (
      <div className={classes}>
        <DualPaneMapper
          handleButtonClick={this.addDatastoreMapping}
          validMapping={
            !(
              selectedTargetDatastore &&
              (selectedSourceDatastores && selectedSourceDatastores.length > 0)
            )
          }
        >
          <DualPaneMapperList
            listTitle="Source Datastores"
            loading={isFetchingSourceDatastores}
          >
            {sourceDatastores &&
              sourceDatastoreFilter(sourceDatastores, input.value).map(item => (
                <DualPaneMapperListItem
                  item={item}
                  text={sourceDatastoreInfo(item)}
                  key={item.id}
                  selected={
                    selectedSourceDatastores &&
                    selectedSourceDatastores.some(
                      sourceDatastore => sourceDatastore.id === item.id
                    )
                  }
                  handleClick={this.selectSourceDatastore}
                  handleKeyPress={this.selectSourceDatastore}
                />
              ))}
            <DualPaneMapperCount
              selectedItems={selectedSourceDatastores.length}
              totalItems={
                sourceDatastoreFilter(sourceDatastores, input.value).length
              }
            />
          </DualPaneMapperList>
          <DualPaneMapperList
            listTitle="Target Datastores"
            loading={isFetchingTargetDatastores}
          >
            {targetDatastores &&
              targetDatastores.map(item => (
                <DualPaneMapperListItem
                  item={item}
                  text={targetDatastoreInfo(item, input.value)}
                  key={item.id}
                  selected={
                    selectedTargetDatastore &&
                    selectedTargetDatastore.id === item.id
                  }
                  handleClick={this.selectTargetDatastore}
                  handleKeyPress={this.selectTargetDatastore}
                />
              ))}
          </DualPaneMapperList>
        </DualPaneMapper>
        <MappingWizardTreeView
          mappings={input.value}
          selectMapping={this.selectMapping}
          removeMapping={this.removeMapping}
          removeAll={this.removeAll}
          selectedMapping={selectedMapping}
        />
      </div>
    );
  }
}

export default DatastoresStepForm;

DatastoresStepForm.propTypes = {
  input: PropTypes.object,
  selectedCluster: PropTypes.object,
  selectedClusterMapping: PropTypes.object,
  sourceDatastores: PropTypes.array,
  targetDatastores: PropTypes.array,
  isFetchingSourceDatastores: PropTypes.bool,
  isFetchingTargetDatastores: PropTypes.bool
};
