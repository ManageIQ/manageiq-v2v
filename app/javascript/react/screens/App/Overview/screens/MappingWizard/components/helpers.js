import { V2V_TARGET_PROVIDERS } from '../MappingWizardConstants';

export const getClusterOptions = clusterMappings => {
  const sourceClustersWithAssociatedTargetClusters = clusterMappings.reduce(
    (mappings, targetClusterWithSourceClusters) => {
      const { nodes: sourceClusters, ...targetCluster } = targetClusterWithSourceClusters;
      const sourceToTargetMappings = sourceClusters.map(sourceCluster => ({
        sourceCluster,
        targetCluster,
        sourceClusterMappedToTargetCluster: {
          name: `${sourceCluster.name} (${targetCluster.name})`,
          id: sourceCluster.id
        }
      }));
      return mappings.concat(sourceToTargetMappings);
    },
    []
  );

  return sourceClustersWithAssociatedTargetClusters.map(
    ({ sourceClusterMappedToTargetCluster }) => sourceClusterMappedToTargetCluster
  );
};

export const multiProviderTargetLabel = (providerType, wizardStep) => {
  switch (providerType) {
    case V2V_TARGET_PROVIDERS[1].id:
      switch (wizardStep) {
        case 'cluster':
          return __('Target Provider \\ Project');
        case 'storage':
          return __('Target Provider \\ Volume Type');
        case 'network':
          return __('Target Project \\ Network');
        default:
          return null;
      }
    default:
      switch (wizardStep) {
        case 'cluster':
          return __('Target Provider \\ Datacenter \\ Cluster');
        case 'storage':
          return __('Target Datastores');
        case 'network':
          return __('Target Network');
        default:
          return null;
      }
  }
};

export const updateMappings = (mappings, clusterStepMappings) =>
  mappings.filter(mapping => {
    const potentialAssociatedClusterStepMappings = clusterStepMappings.filter(
      clusterStepMapping => clusterStepMapping.id === mapping.id
    );

    if (potentialAssociatedClusterStepMappings.length === 0) {
      return false;
    }

    const associatedClusterMapping = potentialAssociatedClusterStepMappings.find(clusterStepMapping => {
      const { nodes: sourceClusters } = clusterStepMapping;
      const sourceClusterIds = sourceClusters.map(sourceCluster => sourceCluster.id);
      const { nodes: mappedSourceStoragesOrNetworks } = mapping.nodes[0];

      return mappedSourceStoragesOrNetworks.every(storageOrNetwork =>
        sourceClusterIds.some(id => id === storageOrNetwork.sourceClusterId)
      );
    });

    return associatedClusterMapping;
  });
