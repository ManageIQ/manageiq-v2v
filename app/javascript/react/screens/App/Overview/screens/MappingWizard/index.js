import { connect } from 'react-redux';
import MappingWizard from './MappingWizard';
import * as MappingWizardActions from './MappingWizardActions';
import {
  mappingWizardOverviewFilter,
  mappingWizardFormFilter
} from './MappingWizardSelectors';

import reducer from './MappingWizardReducer';

export const reducers = { mappingWizard: reducer };

const mapStateToProps = ({ overview, mappingWizard, form }, ownProps) => {
  const selectedOverview = mappingWizardOverviewFilter(overview);
  const selectedForms = mappingWizardFormFilter(form);
  return {
    ...selectedOverview,
    ...selectedForms,
    ...ownProps.data
  };
};

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(mapStateToProps, MappingWizardActions, mergeProps)(
  MappingWizard
);
