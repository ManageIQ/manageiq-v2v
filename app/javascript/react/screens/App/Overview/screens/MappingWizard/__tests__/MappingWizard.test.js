import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MappingWizard from '../MappingWizard';

import { coreComponents } from '../../../../../../../components/';
import componentRegistry from '../../../../../../../components/componentRegistry';

jest.mock('../../../../../../../components/componentRegistry');
componentRegistry.registerMultiple(coreComponents);

describe('MappingWizard component', () => {
  const getBaseProps = () => ({
    hideMappingWizardAction: jest.fn(),
    mappingWizardExitedAction: jest.fn(),
    hideMappingWizard: false
  });

  it('renders the mapping wizard', () => {
    const component = shallow(<MappingWizard {...getBaseProps()} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
