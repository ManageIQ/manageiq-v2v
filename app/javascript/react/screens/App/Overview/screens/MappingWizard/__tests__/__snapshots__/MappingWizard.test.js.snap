// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MappingWizard component renders the mapping wizard 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal-lg wizard-pf"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <Wizard
    className=""
    embedded={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={false}
      closeLabel="Close"
    >
      <button
        aria-hidden="true"
        aria-label="Close"
        className="close"
        onClick={[Function]}
      >
        <Icon
          name="close"
          type="pf"
        />
      </button>
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      >
        Infrastructure Mapping Wizard
      </ModalTitle>
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      className="wizard-pf-body clearfix"
      componentClass="div"
    >
      <MappingWizardBody
        activeStep="1"
        activeStepIndex={0}
        disableNextStep={false}
        goToStep={[Function]}
        hideAlertAction={[Function]}
        loaded={true}
        mappingWizardClustersStep={Object {}}
        mappingWizardDatastoresStep={Object {}}
        mappingWizardGeneralStep={Object {}}
        mappingWizardNetworksStep={Object {}}
        transformationsBody={Object {}}
      />
    </ModalBody>
    <ModalFooter
      bsClass="modal-footer"
      className="wizard-pf-footer"
      componentClass="div"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="btn-cancel"
        disabled={false}
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={true}
        onClick={[Function]}
      >
        <Icon
          name="angle-left"
          type="fa"
        />
        Back
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        disabled={false}
        onClick={[Function]}
      >
        Next
        <Icon
          name="angle-right"
          type="fa"
        />
      </Button>
    </ModalFooter>
  </Wizard>
  <WarningModal
    activeStepIndex={0}
    nextStep={[Function]}
    onFinalStep={false}
  />
</Modal>
`;
